service/src/event/ordering_task.rs

```diff
@@ -1,158 +1,108 @@
 
             tokio::select! {
-                incoming = rx.recv_many(&mut need_prev_buf, 100) => {
-                    if incoming > 0 {
-                        modified = Some(state.add_incoming_batch(need_prev_buf));
-                    }
+                _new = rx_delivered.recv_many(&mut delivered_events, 100) => {
+                    debug!(?delivered_events, "incoming writes!");
                 }
-                new = rx_new.recv_many(&mut newly_added_buf, 100) => {
-                    if new > 0 {
-                        modified = Some(newly_added_buf.into_iter().map(|ev| ev.init_cid).collect::<HashSet<InitCid>>());
+                _new = rx_stream_update.recv_many(&mut updated_streams, 100) => {
+                    debug!(?updated_streams, "incoming updates!");
+                    for stream in updated_streams {
+                        state.add_stream(stream);
                     }
                 }
+                _ = interval.tick() => {
+                    // For this and the else branch, we allow the loop to complete and if state has anything to process, it will be processed.
+                    debug!(stream_count=%state.streams.len(),"Process undelivered stream events interval triggered.");
+                    }
                 else => {
-                    info!(stream_count=%state.pending_by_stream.len(), "Server dropped the ordering task. Processing once more before exiting...");
-                    let _ = state
-                        .process_events(&pool, None)
-                        .await
-                        .map_err(Self::log_error);
-                    return;
+                    debug!(stream_count=%state.streams.len(), "Server dropped the ordering task. Processing once more before exiting...");
                 }
             };
-            // Given the math on OrderingState and the generally low number of updates to streams, we are going
-            // to ignore pruning until there's  more of an indication that it's necessary. Just log some stats.
-            if last_processed.elapsed().as_secs() > CHECK_ALL_INTERVAL_SECONDS {
-                let stream_count = state.pending_by_stream.len();
-                if stream_count > 1000 {
-                    info!(%stream_count, "Over 1000 pending streams without recent updates.");
-                } else {
-                    debug!(%stream_count, "Fewer than 1000 streams pending without recent updates.");
-                }
-            }
 
-            if modified.is_some()
+            if !delivered_events.is_empty()
                 && state
-                    .process_events(&pool, modified)
+                    .process_delivered_events(&pool, delivered_events)
                     .await
                     .map_err(Self::log_error)
                     .is_err()
             {
                 return;
             }
-            last_processed = std::time::Instant::now();
+
+            if state
```

Comment:
looking over every stream every 10 minutes makes me nervous.  I'm worried it is going to paper over bugs in the IOD logic so we never notice them, users will just have occasional sporadic high latency for certain events to be delivered, in a way that will be hard to notice or debug.

It shouldn't be necessary to look at all events, except once at startup.  After that we should be able process streams only in reaction to incoming events, without anything slipping through the cracks.  If there's a bug in our logic, I'd rather us notice it because an event *never* gets delivered, vs the event getting delivered 10 minutes late when this loop happens to notice it.

Reply:
yeah, agreed. the trigger was only reviewing things in memory we failed to process for some reason, but I removed that in favor of explicitly trying anything left over at the beginning of the loop